sensor/kubernetes/crs/crs.go

```diff
@@ -0,0 +1,341 @@
+	}
+
+	centralConnFactory := centralclient.NewCentralConnectionFactory(centralClient)
+	centralConnection := grpcUtil.NewLazyClientConn()
+	certLoader := centralclient.RemoteCertLoader(centralClient)
+	go centralConnFactory.SetCentralConnectionWithRetries(centralConnection, certLoader)
+
+	log.Infof("Connecting to Central server %s", centralEndpoint)
+
+	okSig := centralConnFactory.OkSignal()
+	errSig := centralConnFactory.StopSignal()
+	select {
+	case <-errSig.Done():
+		log.Errorf("failed to get a connection from Central connection factory: %v", errSig.Err())
+		return errors.Wrap(err, "waiting for Central connection from factory")
+	case <-okSig.Done():
+		log.Info("Central connection ready")
+	}
+
+	// New Kubernetes client.
+	config, err := k8sutil.GetK8sInClusterConfig()
+	if err != nil {
+		return errors.Wrap(err, "obtaining in-cluster Kubernetes config")
+	}
+	k8sClient := k8sutil.MustCreateK8sClient(config)
+
+	// Prepare Hello message.
+	deploymentIdentification := sensor.FetchDeploymentIdentification(context.Background(), k8sClient)
+	log.Infof("Determined deployment identification: %s", protoutils.NewWrapper(deploymentIdentification))
+	helmManagedConfigInit, err := getHelmManagedConfig()
+	if err != nil {
+		return errors.Wrap(err, "assembling Helm configuration")
+	}
+	sensorHello := &central.SensorHello{
+		SensorVersion:            version.GetMainVersion(),
+		PolicyVersion:            policyversion.CurrentVersion().String(),
+		DeploymentIdentification: deploymentIdentification,
+	}
+	sensorHello.HelmManagedConfigInit = helmManagedConfigInit
+
+	// Prepare communication channel towards Central.
+	ctx = metadata.AppendToOutgoingContext(ctx, centralsensor.SensorHelloMetadataKey, "true")
+	ctx, err = centralsensor.AppendSensorHelloInfoToOutgoingMetadata(ctx, sensorHello)
+	if err != nil {
+		return errors.Wrap(err, "appending SensorHello to outgoing metadata")
+	}
+	client := central.NewSensorServiceClient(centralConnection)
+	stream, err := communicateWithAutoSensedEncoding(ctx, client)
+	if err != nil {
+		return errors.Wrap(err, "creating central stream with auto-sensed encoding")
+	}
+
+	rawHdr, err := stream.Header()
+	if err != nil {
+		return errors.Wrap(err, "receiving headers from central")
+	}
+
+	hdr := metautils.MD(rawHdr)
+	if hdr.Get(centralsensor.SensorHelloMetadataKey) != "true" {
+		return errors.New("central headers is missing SensorHello metadata key")
+	}
+
+	// Hello Handshake with Central.
+	err = stream.Send(&central.MsgFromSensor{Msg: &central.MsgFromSensor_Hello{Hello: sensorHello}})
+	if err != nil {
+		return errors.Wrap(err, "sending SensorHello message to Central")
+	}
+	log.Debug("Sent SensorHello to Central")
+
+	firstMsg, err := stream.Recv()
+	if err != nil {
+		return errors.Wrap(err, "receiving first message from central")
+	}
+	log.Debug("Received Central response")
+
+	centralHello := firstMsg.GetHello()
+	if centralHello == nil {
```

Comment:
Can we check the central capabilities somewhere here and if it does not support CRS-based registration, log something informative for the user to see? 🤔 
Because an raw message from old central might not be that self-explanatory I guess...

Could you also mention in PR description how the log looks like in the happy and the sad (old central, wrong central) scenario? We should really take some extra time here to make it crystal-clear for the users what's going on, it will pay off in reduced support later.

Reply:
commit is here: 4e390c784e1b8f1c241fdddd88f4156b1b11c972.

Reply:
Will provide the log snippets later.

Reply:
So, I have been looking into the topic of connecting a CRS-enabled sensor to an old central.

The thing is, when you install sensor with CRS enabled, then you need to provide a ... CRS -- a `crs.yaml`. But, this `crs.yaml` will of course need to come from a different (new!) central, which supports CRS. Hence they most likely have a different internal CA, which means that sensor will get some `unknown authority` while trying to connect using the CRS:

```
sensor-7bf94fdc6d-msl9p crs kubernetes/crs: 2024/12/05 09:10:17.188731 crs.go:176: Error: failed to get a connection from Central connection factory: checking central status over HTTP failed: pinging Central: calling https://central.stackrox.svc:443/v1/ping: Get "https://central.stackrox.svc:443/v1/ping": remote error: tls: unknown certificate authority.
sensor-7bf94fdc6d-msl9p crs main: 2024/12/05 09:10:17.188781 main.go:45: Error: Ensuring that this cluster is registered failed: registering secured cluster: opening connection to Central: waiting for Central connection from factory: checking central status over HTTP failed: pinging Central: calling https://central.stackrox.svc:443/v1/ping: Get "https://central.stackrox.svc:443/v1/ping": remote error: tls: unknown certificate authority
```

I will push a commit to get rid of the first log line.

Reply:
Here are the logs for a successful registration:

```
kubernetes/crs: 2024/12/05 12:43:22.849052 crs.go:54: Info: Ensuring certificates for Secured Cluster services are present.
kubernetes/crs: 2024/12/05 12:43:22.849064 crs.go:70: Info: No sensor service certificate found, initiating cluster registration using Cluster Registration Secret.
kubernetes/crs: 2024/12/05 12:43:22.849093 crs.go:91: Info: Trying to load Cluster Registration Secret.
kubernetes/crs: 2024/12/05 12:43:22.849344 crs.go:150: Info: Successfully wrote file /run/secrets/stackrox.io/certs/ca.pem.
kubernetes/crs: 2024/12/05 12:43:22.849402 crs.go:150: Info: Successfully wrote file /run/secrets/stackrox.io/certs/cert.pem.
kubernetes/crs: 2024/12/05 12:43:22.849436 crs.go:150: Info: Successfully wrote file /run/secrets/stackrox.io/certs/key.pem.
kubernetes/crs: 2024/12/05 12:43:22.849676 crs.go:169: Info: Connecting to Central server 34.136.52.95:443
common/centralclient: 2024/12/05 12:43:22.859505 grpc_connection.go:102: Info: Did not add central CA cert to gRPC connection
common/centralclient: 2024/12/05 12:43:22.860977 grpc_connection.go:111: Info: Received max gRPC size from central: 25165824. Overwriting default sensor value of 25165824.
common/centralclient: 2024/12/05 12:43:22.861866 grpc_connection.go:126: Info: Done setting up gRPC connection with central
kubernetes/crs: 2024/12/05 12:43:22.861952 crs.go:180: Info: Central connection ready.
kubernetes/crs: 2024/12/05 12:43:22.891440 crs.go:287: Info: Sensor deployment identification for this secured cluster: {
  "systemNamespaceId": "98424a30-8245-4c44-afa7-df420fbccea7",
  "defaultNamespaceId": "e9564bd3-fc29-4efd-a488-83b53915624e",
  "appNamespace": "stackrox",
  "appNamespaceId": "8d55e705-3cfc-466c-85e5-10ed644dd3ea",
  "appServiceaccountId": "",
  "k8sNodeName": "gke-mc-12-04-mundane-owl-default-pool-855f3085-d14v"
}.
kubernetes/helm: 2024/12/05 12:43:22.891963 helm_managed_config.go:28: Info: Loaded Helm cluster configuration with fingerprint "da329eadf41a2b6c5ae02340642960cd8d363c9ef53be3d946cb040bda731d70"
kubernetes/helm: 2024/12/05 12:43:22.892055 helm_managed_config.go:44: Info: Cluster name from Helm configuration: "foo"
kubernetes/crs: 2024/12/05 12:43:22.936506 crs.go:235: Info: Received CentralHello message from Central cf5a6c71-762c-408f-bb19-4c2f6ad775d8 for Cluster ad2bba7d-7ffc-4451-be0d-f0fecf1872f4.
kubernetes/crs: 2024/12/05 12:43:22.936574 crs.go:253: Info: Persisting retrieved certificates as Kubernetes Secrets in namespace "stackrox".
kubernetes/crs: 2024/12/05 12:43:24.473481 crs.go:271: Info: Successfully persisted received certificates for: SCANNER_SERVICE, SCANNER_DB_SERVICE, COLLECTOR_SERVICE, SENSOR_SERVICE, SCANNER_V4_INDEXER_SERVICE, SCANNER_V4_DB_SERVICE, ADMISSION_CONTROL_SERVICE.
kubernetes/crs: 2024/12/05 12:43:24.473531 crs.go:74: Info: Cluster registration complete.
```

Reply:
There could also be the case where somebody creates a CRS with a new Central, and then rolls back Central to a version that doesn't support CRS. In this case the internal CA would be the same, but the Central wouldn't be able to register the cluster, right?

Reply:
Here are the logs for the scenario described by @vladbologa:

```
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.623794 crs.go:55: Info: Ensuring certificates for Secured Cluster services are present.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.623805 crs.go:71: Info: No sensor service certificate found, initiating cluster registration using Cluster Registration Secret.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.623812 crs.go:92: Info: Trying to load Cluster Registration Secret.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.624017 crs.go:151: Info: Successfully wrote file /run/secrets/stackrox.io/certs/ca.pem.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.624090 crs.go:151: Info: Successfully wrote file /run/secrets/stackrox.io/certs/cert.pem.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.624129 crs.go:151: Info: Successfully wrote file /run/secrets/stackrox.io/certs/key.pem.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.624344 crs.go:170: Info: Connecting to Central server 34.55.190.248:443
sensor-5ff8b675f-45jld crs common/centralclient: 2024/12/10 12:32:36.633503 grpc_connection.go:102: Info: Did not add central CA cert to gRPC connection
sensor-5ff8b675f-45jld crs common/centralclient: 2024/12/10 12:32:36.635408 grpc_connection.go:111: Info: Received max gRPC size from central: 25165824. Overwriting default sensor value of 25165824.
sensor-5ff8b675f-45jld crs common/centralclient: 2024/12/10 12:32:36.636160 grpc_connection.go:126: Info: Done setting up gRPC connection with central
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.636217 crs.go:181: Info: Central connection ready.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.661203 crs.go:297: Info: Sensor deployment identification for this secured cluster: {
sensor-5ff8b675f-45jld crs   "systemNamespaceId": "58e3d020-7996-4583-aaef-7db7c5571884",
sensor-5ff8b675f-45jld crs   "defaultNamespaceId": "19d45a1f-90f3-4412-b7ca-3f9931faa1ac",
sensor-5ff8b675f-45jld crs   "appNamespace": "stackrox",
sensor-5ff8b675f-45jld crs   "appNamespaceId": "e48ac903-5b0b-4a04-957d-682acbcdd368",
sensor-5ff8b675f-45jld crs   "appServiceaccountId": "",
sensor-5ff8b675f-45jld crs   "k8sNodeName": "gke-mc-12-10-past-toad-b-default-pool-dddebad5-fx8q"
sensor-5ff8b675f-45jld crs }.
sensor-5ff8b675f-45jld crs kubernetes/helm: 2024/12/10 12:32:36.661650 helm_managed_config.go:28: Info: Loaded Helm cluster configuration with fingerprint "f96fe8d6aea8800f6fdce7d7212f89c67470d24bb936a6b305ca8221a1ee4ba7"
sensor-5ff8b675f-45jld crs kubernetes/helm: 2024/12/10 12:32:36.661722 helm_managed_config.go:44: Info: Cluster name from Helm configuration: "foo"
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.662767 crs.go:211: Error: Central did not send the SensorHello metadata key after connection attempt using a cluster registration secret.
sensor-5ff8b675f-45jld crs kubernetes/crs: 2024/12/10 12:32:36.662825 crs.go:212: Error: Possible reason: central does not support CRS-based cluster registration.
sensor-5ff8b675f-45jld crs main: 2024/12/10 12:32:36.662874 main.go:45: Error: Ensuring presence of service certificates for this cluster failed: registering secured cluster: handshake with central: central headers are missing the SensorHello metadata key 
```

I would say that looks reasonable.